pdns/recursordist/rec-zonetocache.cc

```diff
@@ -0,0 +1,288 @@
+
+#ifdef HAVE_LIBCURL
+#include "minicurl.hh"
+#endif
+
+#include <fstream>
+
+struct ZoneData
+{
+  ZoneData(shared_ptr<Logr::Logger>& log) :
+    d_log(log) {}
+
+  std::map<pair<DNSName, QType>, vector<DNSRecord>> d_all;
+  std::map<pair<DNSName, QType>, vector<shared_ptr<RRSIGRecordContent>>> d_sigs;
+  std::set<DNSName> d_delegations;
+  time_t d_now;
+  DNSName d_zone;
+  shared_ptr<Logr::Logger>& d_log;
+
+  bool isRRSetAuth(const DNSName& qname, QType qtype);
+  void parseDRForCache(DNSRecord& dr);
+  void getByAXFR(const RecZoneToCache::Config&);
+  void ZoneToCache(const RecZoneToCache::Config& config);
+};
+
+bool ZoneData::isRRSetAuth(const DNSName& qname, QType qtype)
+{
+  DNSName delegatedZone(qname);
+  if (qtype == QType::DS) {
+    delegatedZone.chopOff();
+  }
+  bool isDelegated = false;
+  for (;;) {
+    if (d_delegations.count(delegatedZone) > 0) {
+      isDelegated = true;
+      break;
+    }
+    delegatedZone.chopOff();
+    if (delegatedZone == g_rootdnsname || delegatedZone == d_zone)
+      break;
+  }
+  return !isDelegated;
+}
+
+void ZoneData::parseDRForCache(DNSRecord& dr)
+{
+  const auto key = make_pair(dr.d_name, dr.d_type);
+
+  dr.d_ttl += d_now;
+
+  switch (dr.d_type) {
+  case QType::NSEC:
+  case QType::NSEC3:
+    break;
+  case QType::RRSIG: {
+    const auto& rr = getRR<RRSIGRecordContent>(dr);
+    const auto sigkey = make_pair(key.first, rr->d_type);
+    auto found = d_sigs.find(sigkey);
+    if (found != d_sigs.end()) {
+      found->second.push_back(rr);
+    }
+    else {
+      vector<shared_ptr<RRSIGRecordContent>> sigsrr;
+      sigsrr.push_back(rr);
+      d_sigs.insert({sigkey, sigsrr});
+    }
+    break;
+  }
+  case QType::NS:
+    if (dr.d_name != d_zone) {
+      d_delegations.insert(dr.d_name);
+    }
+    break;
+  default:
+    break;
+  }
+
+  auto found = d_all.find(key);
+  if (found != d_all.end()) {
+    found->second.push_back(dr);
+  }
+  else {
+    vector<DNSRecord> v;
+    v.push_back(dr);
+    d_all.insert({key, v});
+  }
+}
+
+void ZoneData::getByAXFR(const RecZoneToCache::Config& config)
+{
+  ComboAddress primary = ComboAddress(config.d_sources.at(0), 53);
+  uint16_t axfrTimeout = config.d_timeout;
+  size_t maxReceivedBytes = config.d_maxReceivedBytes;
+  const TSIGTriplet tt = config.d_tt;
+  ComboAddress local = config.d_local;
+  if (local == ComboAddress()) {
+    local = pdns::getQueryLocalAddress(primary.sin4.sin_family, 0);
+  }
+
+  AXFRRetriever axfr(primary, d_zone, tt, &local, maxReceivedBytes, axfrTimeout);
+  Resolver::res_t nop;
+  vector<DNSRecord> chunk;
+  time_t axfrStart = time(nullptr);
+  time_t axfrNow = time(nullptr);
+
+  while (axfr.getChunk(nop, &chunk, (axfrStart + axfrTimeout - axfrNow))) {
+    for (auto& dr : chunk) {
+      parseDRForCache(dr);
+    }
+    axfrNow = time(nullptr);
+    if (axfrNow < axfrStart || axfrNow - axfrStart > axfrTimeout) {
+      throw PDNSException("Total AXFR time for zoneToCache exceeded!");
```

Comment:
Is there a particular reason you are using `PDNSException` and not an exception derived from `std::exception` instead? I would prefer less custom exceptions whenever possible :)

Reply:
This is a copy-paste from rpzloader